### ISSUE TITLE
Change link API to allow validation to be deferred until a modifier is pressed

### ISSUE THREAD

**Issue by Tyriar** [xtermjs/xterm.js#1825]
@nojvek proposed this idea to me:

Instead of the current link API (still marked experimental) where all links are parsed and validated when the viewport stops changing, it could be changed to parse and validate only when the viewport has stopped _and_ some modifier is pressed (eg. ctrl or cmd, defined by embedder). That way a lot of file exists checks could be deferred until when we want them. Plus there is still likely a delay between pressing the modifier and mousing over the link, so the user probably won't be able to tell the difference.

**Comment by jerch** [comment]
Great idea - it would also lift some burden from input flow, esp when there are complex matcher, and many of them. :+1:
Only possible downside I can imagine - if an integrator relies on the instant linkification, this wont work anymore. But this can still be achieved by doing the linkification explicitly. So no showstopper imho.

**Comment by Tyriar** [comment]
It would require modifiers yes, right now by default links will not require any modifier. Not even sure anyone is using it without hooking up a modifier?

**Comment by mofux** [comment]
IMO, I'd also prefer to only follow links when pressing a modfier, but linkification should start right away (as in a link is visible right at the start, but follow requires consent). I think having `cmd+click` would be a good default for `MacOS`, while `ctrl+click` would be a good default for other platforms (which should follow vscode behaviour).  Not sure about showing a (localized / customisable?) tooltip?

**Comment by Tyriar** [comment]
I guess doing this would remove underline/tooltip on hover (which vscode does have). The problem is the validation is the thing that times time as it needs to go to the disk...

**Comment by mofux** [comment]
Yeah you're right, at some point I'd really like to have linkification be done on the particular part of the buffer as it arrives (tracking it as some kind of decoration), so we don't have to evaluate links over and over again as they move through the viewport. This would solve the problem of overscanning the viewport for links with every scroll. On the other hand it would certainly hurt the runtime performance, especially if lots of data is arriving (`seq 1000000` )...

**Comment by mofux** [comment]
...acutally, that feature would be a good reference implementation for the Marker API, marking and referencing a link as long as it persists:
- re-evaluate the link if the buffer line(s) get modified / removed while still being part of the "hot" viewport (as in not being in the scrollback buffer yet)
- disposing the link as it leaves the scrollback

**Comment by Tyriar** [comment]
Remembering links using markers sounds like a good idea, just need to invalidate them if that line changes 😃
